app/src/main/java/org/mozilla/focus/locale/screen/LocaleDescriptor.kt

```diff
@@ -0,0 +1,152 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.locale.screen
+
+import android.text.TextUtils
+import android.util.Log
+import org.mozilla.focus.locale.Locales
+import java.text.Collator
+import java.util.Locale
+
+class LocaleDescriptor(private val localeTag: String) : Comparable<LocaleDescriptor> {
+
+    private val languageCodeAndNameMap: HashMap<String, String> = HashMap()
+    private var nativeName: String? = null
+
+    init {
+        fillLanguageCodeAndNameMap()
+        setupLocaleDescriptor()
+    }
+
+    private fun fillLanguageCodeAndNameMap() {
+        // Only ICU 57 actually contains the Asturian name for Asturian, even Android 7.1 is still
+        // shipping with ICU 56, so we need to override the Asturian name (otherwise displayName will
+        // be the current locales version of Asturian, see:
+        // https://github.com/mozilla-mobile/focus-android/issues/634#issuecomment-303886118
+        languageCodeAndNameMap["ast"] = "Asturianu"
+        // On an Android 8.0 device those languages are not known and we need to add the names
+        // manually. Loading the resources at runtime works without problems though.
+        languageCodeAndNameMap["cak"] = "Kaqchikel"
+        languageCodeAndNameMap["ia"] = "Interlingua"
+        languageCodeAndNameMap["meh"] = "Tu´un savi ñuu Yasi'í Yuku Iti"
+        languageCodeAndNameMap["mix"] = "Tu'un savi"
+        languageCodeAndNameMap["trs"] = "Triqui"
+        languageCodeAndNameMap["zam"] = "DíɁztè"
+        languageCodeAndNameMap["oc"] = "occitan"
+        languageCodeAndNameMap["an"] = "Aragonés"
+        languageCodeAndNameMap["tt"] = "татарча"
+        languageCodeAndNameMap["wo"] = "Wolof"
+        languageCodeAndNameMap["anp"] = "अंगिका"
+        languageCodeAndNameMap["ixl"] = "Ixil"
+        languageCodeAndNameMap["pai"] = "Paa ipai"
+        languageCodeAndNameMap["quy"] = "Chanka Qhichwa"
+        languageCodeAndNameMap["ay"] = "Aimara"
+        languageCodeAndNameMap["quc"] = "K'iche'"
+        languageCodeAndNameMap["tsz"] = "P'urhepecha"
+        languageCodeAndNameMap["jv"] = "Basa Jawa"
+        languageCodeAndNameMap["ppl"] = "Náhuat Pipil"
+        languageCodeAndNameMap["su"] = "Basa Sunda"
+        languageCodeAndNameMap["hus"] = "Tének"
+        languageCodeAndNameMap["co"] = "Corsu"
+        languageCodeAndNameMap["sn"] = "ChiShona"
+        languageCodeAndNameMap["zh-CN"] = "中文 (中国大陆)"
```

Comment:
(Off-topic): Eventually we should have this map somewhere in A-C, to avoid having this in Focus and Fenix.

Reply:
We actually do this quite differently in Fenix. https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleManagerExtension.kt#16

Reply:
We still seem to have a map though. For some reason containing a lot of strings that should be in Android already though.
https://searchfox.org/mozilla-mobile/source/fenix/app/src/main/java/org/mozilla/fenix/settings/advanced/LocaleViewHolders.kt#67